lib/services/local/vnet_config.go

```diff
@@ -0,0 +1,110 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package local
+
+import (
+	"context"
+	"log/slog"
+	"net"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/gen/proto/go/teleport/vnet/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/services/local/generic"
+	"github.com/gravitational/trace"
+)
+
+const (
+	vnetConfigPrefix        = "vnet_config"
+	vnetConfigSingletonName = "vnet-config"
+)
+
+type VnetConfigService struct {
+	slog *slog.Logger
+	svc  *generic.ServiceWrapper[*vnet.VnetConfig]
+}
+
+func NewVnetConfigService(backend backend.Backend) (*VnetConfigService, error) {
+	svc, err := generic.NewServiceWrapper(
+		backend,
+		types.KindVnetConfig,
+		vnetConfigPrefix,
+		services.MarshalProtoResource[*vnet.VnetConfig],
+		services.UnmarshalProtoResource[*vnet.VnetConfig],
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &VnetConfigService{
+		svc:  svc,
+		slog: slog.With(teleport.ComponentKey, "VnetConfig.local"),
+	}, nil
+}
+
+func (s *VnetConfigService) GetVnetConfig(ctx context.Context) (*vnet.VnetConfig, error) {
+	return s.svc.GetResource(ctx, vnetConfigSingletonName)
+}
+
+func (s *VnetConfigService) CreateVnetConfig(ctx context.Context, vnetConfig *vnet.VnetConfig) (*vnet.VnetConfig, error) {
+	if err := validateVnetConfig(vnetConfig); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return s.svc.CreateResource(ctx, vnetConfig)
+}
+
+func (s *VnetConfigService) UpdateVnetConfig(ctx context.Context, vnetConfig *vnet.VnetConfig) (*vnet.VnetConfig, error) {
+	if err := validateVnetConfig(vnetConfig); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return s.svc.ConditionalUpdateResource(ctx, vnetConfig)
+}
+
+func (s *VnetConfigService) UpsertVnetConfig(ctx context.Context, vnetConfig *vnet.VnetConfig) (*vnet.VnetConfig, error) {
+	if err := validateVnetConfig(vnetConfig); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return s.svc.UpsertResource(ctx, vnetConfig)
+}
+
+func (s *VnetConfigService) DeleteVnetConfig(ctx context.Context) error {
+	return s.svc.DeleteResource(ctx, vnetConfigSingletonName)
+}
+
+func validateVnetConfig(vnetConfig *vnet.VnetConfig) error {
+	if vnetConfig.GetKind() != types.KindVnetConfig {
+		return trace.BadParameter("kind must be %q", types.KindVnetConfig)
+	}
+	if vnetConfig.GetVersion() != types.V1 {
+		return trace.BadParameter("version must be %q", types.V1)
+	}
+	if vnetConfig.GetMetadata().GetName() != vnetConfigSingletonName {
+		return trace.BadParameter("name must be %q", vnetConfigSingletonName)
+	}
+	if cidrRange := vnetConfig.GetSpec().GetIpv4CidrRange(); cidrRange != "" {
```

Comment:
If not set, what's the default cidr range used? 
Should we handle that case an explicitly set it in the vnet config for people to understand?

Reply:
where is the default set?

Reply:
the default doesn't need to actually be in the `vnet_config` resource. "clients" will use the default if a `vnet_config` either does not exist or does not have `ipv4_cidr_range` set.

Reply:
makes sense, good point about terraform / kube operator too

Reply:
>  don't things like terraform get mad when the created resource doesn't match what was given?

@nklaassen 

Yes,  but we have a process for all our resources that allows circumvent that limitation. Terraform calls `CheckAndSetDefaults` and copies fields before creating the resource. 
I would say this makes sense to identify the cidr by default but I leave the decision to you